src/core/fonts.js

```diff
@@ -3351,9 +3352,119 @@ var Font = (function FontClosure() {
       return (charsCache[charsCacheKey] = glyphs);
     },
 
+    getGlyphsAndCharPosition: function Font_getGlyphsAndCharPosition(chars) {
+      const glyphs = [];
+
+      if (this.cMap) {
+        const c = Object.create(null);
+        let i = 0;
+        while (i < chars.length) {
+          this.cMap.readCharCode(chars, i, c);
+          const length = c.length;
+          const isSpace = length === 1 && chars.charCodeAt(i) === 0x20;
+          const glyph = this._charToGlyph(c.charcode, isSpace);
+          glyphs.push({
+            start: i,
+            end: i + length,
+            glyph,
+          });
+          i += length;
+        }
+      } else {
+        for (let i = 0, ii = chars.length; i < ii; ++i) {
+          const charcode = chars.charCodeAt(i);
+          const glyph = this._charToGlyph(charcode, charcode === 0x20);
+          glyphs.push({
+            start: i,
+            end: i + 1,
+            glyph,
+          });
+        }
+      }
+
+      return glyphs;
+    },
+
+    getCharPosition: function Font_getCharPosition(chars) {
+      // Chars can have different sizes
+      // so get the char positions in the string
+      // it's useful to split a string at correct char positions.
+      const positions = [];
+
+      if (this.cMap) {
+        const c = Object.create(null);
+        let i = 0;
+        while (i < chars.length) {
+          this.cMap.readCharCode(chars, i, c);
+          const length = c.length;
+          positions.push([i, i + length]);
+          i += length;
+        }
+      } else {
+        for (let i = 0, ii = chars.length; i < ii; ++i) {
+          positions.push([i, i + 1]);
+        }
+      }
+
+      return positions;
+    },
+
     get glyphCacheValues() {
       return Object.values(this.glyphCache);
     },
+
+    encodeString: function Font_encodeString(str) {
+      // The function returns an array of strings
+      // at even positions there is a valid string (may be empty)
+      // and at odd positions string where chars are not in the font
+      const buffers = [];
+      let buf = [];
+      let isError = false;
+
+      for (let i = 0, ii = str.length; i < ii; i++) {
+        const unicode = str.codePointAt(i);
+        let hasError = false;
+        if (unicode > 0xd7ff && (unicode < 0xe000 || unicode > 0xfffd)) {
+          // unicode is represented by two u16
```

Comment:
What is `u16`? Please improve/expand this comment, since it's not all that clear currently!